webrtc/RTCPeerConnection-addIceCandidate.html

```diff
@@ -0,0 +1,531 @@
+
+  function assert_never_resolves(t, promise) {
+    promise.then(
+      t.step_func(result => {
+        assert_unreached(`Pending promise should never be resolved. Instead it is fulfilled with: ${result}`);
+      }),
+      t.step_func(err => {
+        assert_unreached(`Pending promise should never be resolved. Instead it is rejected with: ${err}`);
+      }));
+
+    t.step_timeout(t.step_func_done(), 100);
+  }
+
+  async_test(t => {
+    const pc = new RTCPeerConnection();
+
+    pc.setRemoteDescription(sessionDesc)
+    .then(() => {
+      assert_never_resolves(t,
+        pc.addIceCandidate({
+          candidate: candidateStr1,
+          sdpMid, sdpMLineIndex, ufrag
+        }));
+
+      pc.close();
+
+      // When pc is closed, the remote description is not modified
+      // even if succeed
+      t.step_timeout(t.step_func(() => {
+        assert_false(pc.remoteDescription.sdp.includes(candidateLine1),
+          'Candidate should not be added to SDP because pc is closed');
+      }), 80);
+    });
+  }, 'Add valid candidate should never resolve when pc is closed');
+
+  async_test(t => {
+    const pc = new RTCPeerConnection();
+
+    assert_never_resolves(t, pc.addIceCandidate({
+      candidate: candidateStr1,
+      sdpMid, sdpMLineIndex, ufrag
+    }));
+
+    pc.close();
+  }, 'Add candidate when remote description is null should never resolve when pc is closed');
+
+  async_test(t => {
+    const pc = new RTCPeerConnection();
+
+    pc.setRemoteDescription(sessionDesc)
+    .then(() => {
+      assert_never_resolves(t, pc.addIceCandidate({
+        candidate: invalidCandidateStr,
+        sdpMid, sdpMLineIndex, ufrag
+      }));
+      pc.close();
+    });
+  }, 'Add candidate with invalid candidate string should never resolve when pc is closed');
+
+  async_test(t => {
+    const pc = new RTCPeerConnection();
+
+    pc.setRemoteDescription(sessionDesc)
+    .then(() => {
+      assert_never_resolves(t, pc.addIceCandidate({
+        candidate: candidateStr1,
+        sdpMid: 'invalid',
+        sdpMLineIndex, ufrag
+      }));
+      pc.close();
+    });
+  }, 'Add candidate with invalid sdpMid should never resolve when pc is closed');
+
+  async_test(t => {
+    const pc = new RTCPeerConnection();
+
+    pc.setRemoteDescription(sessionDesc)
+    .then(() => {
+      assert_never_resolves(t, pc.addIceCandidate({
+        candidate: candidateStr1,
+        sdpMLineIndex: 2,
+        ufrag
+      }));
+      pc.close();
+    });
+  }, 'Add candidate with invalid sdpMLineIndex should never resolve when pc is closed');
+
+  async_test(t => {
+    const pc = new RTCPeerConnection();
+
+    pc.setRemoteDescription(sessionDesc)
+    .then(() => {
+      assert_never_resolves(t, pc.addIceCandidate({
+        candidate: invalidCandidateStr,
+        sdpMid, sdpMLineIndex, ufrag
+      }));
+      pc.close();
+    });
+  }, 'Add candidate with invalid candidate string should never resolve when pc is closed');
```

Comment:
Hmm I must have made one extra copy when duplicating and editing the tests. Will remove.

I think testharness doesn't report it because it report the timeout error instead for _Add candidate with both sdpMid and sdpMLineIndex null should still reject with TypeError after pc is closed_.